custom/icds_reports/utils/aggregation_helpers/distributed/adolescent_girls_registration.py

```diff
@@ -0,0 +1,49 @@
+from dateutil.relativedelta import relativedelta
+
+from custom.icds_reports.const import AGG_ADOLESCENT_GIRLS_REGISTRATION_TABLE
+from custom.icds_reports.utils.aggregation_helpers import month_formatter
+from custom.icds_reports.utils.aggregation_helpers.distributed.base import (
+    StateBasedAggregationDistributedHelper,
+)
+
+
+class AggAdolescentGirlsRegistrationAggregate(StateBasedAggregationDistributedHelper):
+    helper_key = 'adolescent-girls'
+    ucr_data_source_id = 'static-adolescent_girls_reg_form'
+    aggregate_parent_table = AGG_ADOLESCENT_GIRLS_REGISTRATION_TABLE
+
+    def aggregation_query(self):
+        month = self.month.replace(day=1)
+        current_month_start = month_formatter(self.month)
+        next_month_start = month_formatter(self.month + relativedelta(months=1))
+
+        query_params = {
+            "month": month_formatter(month),
+            "state_id": self.state_id,
+            "current_month_start": current_month_start,
+            "next_month_start": next_month_start,
+        }
+
+        return """
+        INSERT INTO "{tablename}" (
+          state_id, supervisor_id, awc_id, month, person_case_id, out_of_school, 
+          re_out_of_school, admitted_in_school
+        ) (
+        SELECT DISTINCT ON (person_case_id)
+            %(state_id)s AS state_id,
+            supervisor_id,
+            awc_id,
+            %(month)s::DATE AS month,
+            person_case_id,
+            out_of_school IS NOT NULL AND out_of_school='yes' AS out_of_school,
+            re_out_of_school IS NOT NULL AND re_out_of_school='yes' AS re_out_of_school,
+            admitted_in_school IS NOT NULL AND admitted_in_school='yes' AS admitted_in_school
+            from "{ucr_tablename}"
+            WHERE state_id = %(state_id)s AND
+                timeend < %(next_month_start)s
```

Comment:
This basically needs to have condition of last form submitted ever. I could have rollover from previous month. But for the first time that would fail. 
Since the query might be quite big So, will be thinking about the optimization probably limiting the rows being read from ucr.

Reply:
Why would rolling over fail for the first time? If you do an OUTER JOIN shouldn't those records be included?

For example: https://github.com/dimagi/commcare-hq/blob/6a2883f71448655869612ebcb2866c591ee10bdd/custom/icds_reports/utils/aggregation_helpers/distributed/growth_monitoring_forms.py#L171-L176

That uses a lot of COALESCE and GREATEST to accomplish that without reading all forms. (Of course if this UCR is only going to have a few forms this may be ok. I have no idea how common this form will be)

Reply:
Hey, Sorry by failing I meant, it wont fail syntactically. I mean the evaulated indicator would be wrong because in the query from ucr I would consider just this months form and full join with previous month table And is there wont be any previous month data. It the evaulated column would have built with incomplete forms.  The indicator he such, we need to consider last ever submitted and make the indicator based on that.

Reply:
by the way. what is the  upper limit to which query all forms is fine?

Reply:
Hey @emord  checked with GS team. Their estimation is atleast next 1 year, the number of forms wont increase 400K per state? Do you think its just fine to look all forms in that case?

Reply:
That's a fairly small number of forms and this query could probably complete quickly, but eventually it will introduce more disk i/o which has been a real killer for the agg and databases in general, so it will introduce more contention for concurrent running queries.

> there wont be any previous month data

When I first began adding this type of table, I would often run the migration for the table and then run something like:

```python
for month in months_from_2015_to_now:
    for state_id in state_ids:
        AdolescentGirlAgg.aggregate(month, state_id)
```

which would create the previous tables and consider all the data from the past and allow the agg to focus only on the current month's data

Reply:
Aaah OK. I had thought of that But was not sure how right would be that. But This change will be under feature flag and under QA for some time so we should get time to run this.
Cool. Making this change. Thanks

Reply:
@emord  made the change. Changed the query to use rollover from previous month